src/util.rs

```diff
@@ -55,3 +55,329 @@ impl Row for Index {
 pub fn offsets(bit: Index) -> (usize, usize, usize) {
     (bit.offset(SHIFT1), bit.offset(SHIFT2), bit.offset(SHIFT3))
 }
+
+/// Finds the highest bit that splits set bits of the `usize`
+/// to half (rounding up).
+///
+/// Returns `None` if the `usize` has only one or zero set bits.
+///
+/// # Examples
+/// ````rust,ignore
+/// use hibitset::util::average_ones;
+///
+/// assert_eq!(Some(3), average_ones(0b10110));
+/// assert_eq!(Some(6), average_ones(0b100010));
+/// assert_eq!(None, average_ones(0));
+/// assert_eq!(None, average_ones(1));
+/// ````
+// TODO: Can 64/32 bit variants be merged to one implementation?
+pub fn average_ones(n: usize) -> Option<usize> {
+    #[cfg(target_pointer_width= "64")]
+    fn average(n: usize) -> Option<usize> {
+        average_ones_u64(n as u64).map(|n| n as usize)
+    }
+    #[cfg(target_pointer_width= "32")]
+    fn average(n: usize) -> Option<usize> {
+        average_ones_u64(n as u32).map(|n| n as usize)
+    }
+    average(n)
+}
+
+#[allow(dead_code)]
+fn average_ones_u32(n: u32) -> Option<u32> {
+    use std::num::Wrapping as W;
+    let n = W(n);
+
+    // !0 / ((1 << (1 << n)) | 1)
+    const PAR: [W<u32>; 5] = [
+        W(!0 / 0x3),
+        W(!0 / 0x5),
+        W(!0 / 0x11),
+        W(!0 / 0x101),
+        W(!0 / 0x10001)
+    ];
+
+    // Counting set bits in parallel
+    let a = n - ((n >> 1) & PAR[0]);
```

Comment:
You can just use the `count_ones` method here

Reply:
I can't use it as I need the partial results of the counting for the binary search. If this wasn't the case I would have used it in a heart beat.

Reply:
Ah okay, makes sense.